cohttp-eio/src/parser.ml

```diff
@@ -0,0 +1,207 @@
+open Angstrom
+
+let token =
+  take_while1 (function
+    | '0' .. '9'
+    | 'a' .. 'z'
+    | 'A' .. 'Z'
+    | '!' | '#' | '$' | '%' | '&' | '\'' | '*' | '+' | '-' | '.' | '^' | '_'
+    | '`' | '|' | '~' ->
+        true
+    | _ -> false)
+
+let space = char '\x20'
+let htab = char '\t'
+let ows = skip_many (space <|> htab)
+let optional x = option None (x >>| Option.some)
+let vchar = satisfy (function '\x21' .. '\x7E' -> true | _ -> false)
+let digit = satisfy (function '0' .. '9' -> true | _ -> false)
+let crlf = string_ci "\r\n" <?> "[crlf]"
+
+(*-- https://datatracker.ietf.org/doc/html/rfc7230#section-3.2 --*)
+let headers =
+  let header_field =
+    let* header_name = token <* char ':' <* ows >>| String.lowercase_ascii in
+    let+ header_value =
+      let field_content =
+        let c2 =
+          optional
+            (let+ c1 = skip_many1 (space <|> htab) *> vchar in
+             Format.sprintf " %c" c1)
+          >>| function
+          | Some s -> s
+          | None -> ""
+        in
+        lift2 (fun c1 c2 -> Format.sprintf "%c%s" c1 c2) vchar c2
```

Comment:
We did a number of benchmarks some time ago. Using `String.concat` or `Buffer` can be orders of magnitude faster than `Printf.sprintf`. Of course, here we are using `%X`, so we may not have a an alternative without relying on extra code or library.

I still think that you should try to use our parser from http, where needed we can adjust it or abstract it over bytes/bigstrings. It feels wrong to ship two different http parsers.

Reply:
Let me get back to you on this issue a bit later.

Reply:
@mseri `cohttp-eio` now uses the parser included in `http` package.

Reply:
🥳 thanks! Did you notice any problem with it performance-wise?

Reply:
Performance was good, slightly better than with angstrom based parser. Do you think if using bigstring as the base type in parser(instead of string) and thus eliminating bigstring to string conversion cost will increase in further performance boost?

Reply:
I just did a multi-core benchmark of the `http` parser against `httpaf-eio` and `nethttp-go`, It seems the  parser doesn't scale linearly as compared to the other two. It starts plateauing in/around 300K req/second.
![image](https://user-images.githubusercontent.com/315907/154101719-58f9c43b-8fb5-4a6e-aab2-3c1d0481edbe.png)
The branch is here if your are interested. https://github.com/bikallem/ocaml-cohttp/blob/eio-3-multi-core/cohttp-eio/src/reader.ml

Reply:
It seems `eio_linux` with polling plateaus around 300k, so I changed it back to `eio_main`. With that change we don't have any scaling issue. The latest multicore benchmark (48 cores) is a below. It scales very nicely without ever plateauing. 
![image](https://user-images.githubusercontent.com/315907/154302187-6529f2e4-08f7-487f-880d-cee7a5965387.png)

Reply:
Wow. Impressive

Reply:
> The latest multicore benchmark (48 cores) is below. It scales very nicely without ever plateauing.

Perhaps this is true, but the graph doesn't show that:
- With 1 core, it shows it's possible to do about 100,000 requests per second
- With 48 cores it shows the new code doing about 400,000 requests per second

We don't know how fast it goes with 48 cores because the x-axis only goes up to 400,000 and all the multicore versions can handle that with no problem, as expected. We need to know where each one levels off.

It would be more useful to graph the maximum speed against the number of cores being used. Possibly replacing `wrk2` with `wrk` would make that easier.